### PR TITLE
Parser: add recovery for unfinished match clauses

### PR DESCRIPTION
**Pull request by auduchinok**
The following examples will now produce match clauses (previously it'd fail and skip them):

```fsharp
match () with
| x
```

```fsharp
match () with
| x when e
| otherClause -> ()
```

The first case recovery and diagnostic improvements were done by @dsyme during our parser recovery session.

### CODE REVIEW

**Comment by KevinRansom** [comment]
@auduchinok , perhaps you could resolve the conflicts, and could we also have some kind of test.

Kevin

**Comment by auduchinok** [comment]
I'm not able to reproduce the desired behaviour in the remaining test, `VariableIdentifier.AsParameter`: I do get the completion items in both `main` and this branch when running a debug IDE instance, and it fails to get the items when running the test.

I am also unable to debug the test in neither Rider nor VS 2019, since there're some different build issues in both IDEs on my machine.

File: src/fsharp/pars.fsy [hunk 1]
```diff
@@ -3123,10 +3146,12 @@ constrPattern:
 
 atomicPatsOrNamePatPairs:
   | LPAREN namePatPairs rparen
-      { SynArgPats.NamePatPairs $2 }
+      { SynArgPats.NamePatPairs $2, snd $2 }
 
   | atomicPatterns
-      { SynArgPats.Pats $1 }
+      { let mWhole = rhs parseState 1
+        let m = (mWhole.StartRange, $1) ||> unionRangeWithListBy (fun p -> p.Range)
```
Comment (dsyme) [inline src/fsharp/pars.fsy:3153]: maybe call this `mAll`

File: tests/service/ParserTests.fs [hunk 2]
```diff
@@ -31,16 +30,50 @@ type U2 =
...
+"""
+
+    match getSingleExprInModule parseResults with
+    | SynExpr.Match (_, _, [ SynMatchClause (_, _, _, SynExpr.ArbitraryAfterError _, _, _) ], _) -> ()
+    | _ -> failwith "Unexpected tree"
+
+[<Test>]
+let ``Match clause 03 - When`` () =
+    let parseResults = getParseResults """
+match () with
```
Comment (dsyme) [inline tests/service/ParserTests.fs:72]: Could you add a test for this, assuming it works?
```fsharp
match () with
| x 
| _ -> ()
```

Comment (dsyme) [reply tests/service/ParserTests.fs:72]: Also see if this recovers:
```fsharp
match () with
| x ->
| _ -> ()
```

File: tests/service/Common.fs [hunk 3]
```diff
@@ -253,6 +253,19 @@ let getSingleModuleLikeDecl (input: ParsedInput) =
     | ParsedInput.ImplFile (ParsedImplFileInput (modules = [ decl ])) -> decl
     | _ -> failwith "Could not get module decls"
 
+let getSingleModuleMemberDecls (input: ParsedInput) =
+    match getSingleModuleLikeDecl input with
+    | SynModuleOrNamespace (decls = decls) -> decls
+
+let getSingleExprInModule (input: ParsedInput) =
+    match getSingleModuleLikeDecl input with
+    | SynModuleOrNamespace (decls = decls) ->
```
Comment (dsyme) [inline tests/service/Common.fs:262]: Would normally use
```fsharp
let (SynModuleOrNamespace (decls = decls)) = getSingleModuleLikeDecl input
```

Comment (dsyme) [reply tests/service/Common.fs:262]: Why not call `getSingleModuleMemberDecls` here?

**Review by dsyme — APPROVED** [review]
Added some comments

**Comment by dsyme** [comment]
> I'm not able to reproduce the desired behaviour in the remaining test, VariableIdentifier.AsParameter: I do get the completion items in both main and this branch when running a debug IDE instance, and it fails to get the items when running the test.

Given the validation steps you did then we could consider disabling the test.  I'll try to reproduce.

File: tests/service/Common.fs (hunk 3 above)
Comment (auduchinok) [reply tests/service/Common.fs:262]: > Why not call getSingleModuleMemberDecls here?

I think that's what I originally had wanted and didn't finish while were debugging the tests. Thanks, I'll update it.

File: tests/service/ParserTests.fs (hunk 2 above)
Comment (auduchinok) [reply tests/service/ParserTests.fs:72]: Added both cases, both are green. 🙂

**Comment by auduchinok** [comment]
Yay, all is green, sans the disabled test.

**Comment by dsyme** [comment]
Locally I just got this:

> C:\GitHub\dsyme\fsharp>.\build -c Release -testFSharpQA
> 
> 0 tests failed (0.00%)

So it is confusing.  The test does however relate to the area that you're changing.  Hmmm...

File: vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs [hunk 4]
```diff
@@ -6239,7 +6239,7 @@ let rec f l =
             marker = "(*MarkerMethodInType*)",
             list = ["PrivateMethod"])       
 
-    [<Test>]
+//    [<Test>]
```
Comment (dsyme) [inline vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs:6242]: Best to use `Disabled` or `Skip` or whatever this test quite uses

**Comment by dsyme** [comment]
I will merge this, partly because the error recovery implemented is far more useful that the very corner-case test (which may not be a true failure since we can't repro it locally)

File: vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs (hunk 4 above)
Comment (auduchinok) [reply vsintegration/tests/UnitTests/LegacyLanguageService/Tests.LanguageService.Completion.fs:6242]: Sorry, I should've tried to use some attribute indeed. I looked at the code around and used the same approach.

**Comment by auduchinok** [comment]
> Locally I just got this on this branch

It was the VS test suite, not the FSharpQA one. Could you try that one?

**Pull request merged**